issue-1132 - refactor multi tenancy and schema migration docs (#1165)
IBM/FHIR: fhir-persistence-schema/docs/SchemaMigration.md

| PostgreSQL |        12 | Experimental. Single tenant per database. |
| Derby      | 10.14.2.0 | Development only. Single tenant per database |

----------------------------------------------------------------
# Overview and Concepts

The IBM FHIR schema is managed in code and is applied to a target database using a JDBC connection. This simplifies development of Derby-based unit-tests, allows a common schema definition to be used across multiple database flavors, simplifies deployment in cloud scenarios, and because the data model dependencies are understood by the code, the schema creation process can be parallelized to reduce deployment times. In practice, however, we've found that the number of threads must be limited due to driving contention in the internal catalog tables (in DB2, for example). We have also found that some parallel object create operations cause internal database deadlocks (notably creating foreign key relationships). The implementation contains a retry loop, but may fail if the retry limit is exceeded. Note that these deadlocks are internal to Db2, not the more common kind typically caused by poor application code.

> The `FHIR_ADMIN` schema is used to manage meta-data related to the actual data schemas. A single database instance can be used to support multiple FHIR data schemas, as long as the FHIR_ADMIN schema structure doesn't change. At the time of writing, the IBM FHIR Server does not support more than one FHIR_ADMIN schema in a single database. If a use-case arises where this is required, it is simply a matter of identifying in code where FHIR_ADMIN is used as a constant and replacing instances with a configurable property, although an implementation may also want to consider protecting individual schemas from accidentally being managed by more than one FHIR_ADMIN schema. One possible use-case for this sharing is using a database instance for schema development work, although use of the Db2 docker container makes this moot.

Reviewer: ```suggestion
The `FHIR_ADMIN` schema is used to manage meta-data related to the actual data schemas. A single database instance can be used to support multiple FHIR data schemas, as long as the FHIR_ADMIN schema structure doesn't change. At the time of writing, the IBM FHIR Server does not support more than one FHIR_ADMIN schema in a single database. If a use-case arises where this is required, it is simply a matter of identifying in code where FHIR_ADMIN is used as a constant and replacing instances with a configurable property, although an implementation may also want to consider protecting individual schemas from accidentally being managed by more than one FHIR_ADMIN schema. One possible use-case for this sharing is using a database instance for schema development work.
```

Author: ok

The `FHIR_ADMIN` schema also plays an important role in managing tenants in the Db2 multi-tenant implementation. See the [Db2 Multi-Tenant Schema Design](DB2MultiTenancy.md) for more details.

## Differences Among Database Implementations

The following table highlights the main differences among the database implementations currently supported:

| Database   | Variation |
| --------   | ----------|
| DB2        | Multi-tenant. Supports multiple but isolated tenants within one database/schema |
| DB2        | An SPL stored procedure is used to implement the resource storage logic, reducing the number of application server to database round-trips, improving performance |
| DB2        | Uses `FHIR_TS` tablespace for admin tables, and a tablespace per tenant |
| DB2        | `FHIR_TS` is created using a small extent size for efficiency |

> | PostgreSQL | Based on the Derby implementation, but uses a stored procedure for the resource persistence logic, like DB2 |

Reviewer: ```suggestion
| PostgreSQL | Based on the Derby implementation, but uses a function for the resource persistence logic, like DB2 |
```

Author: OK

The DDL for most objects (like tables) is specified once. Changes to the table structure are applied as alter statements and assigned an increasing version id. We use a simple version number tracking to identify which changes (deltas) need to be applied to a particular object.

The schema update utility first reads the VERSION_HISTORY table, loading all records for the target schema (e.g. FHIRDATA). The utility only applies changes which have a version number greater than the currently recorded version. Once the DDL has been applied successfully, the version number is updated in VERSION_HISTORY. This makes the processed idempotent. Subsequent runs of the schema update utility only apply changes which have a greater version id value than the most recently stored value for each object.

Enhancements to schema migration handling are tracked in [issue-1163](https://github.com/IBM/FHIR/issues/1163).

> 

Reviewer: let's not cite these, it's hard to back track

Author: ok

## Schema: FHIR_ADMIN

The Schema FHIR_ADMIN is deployed once per database and contains tables used to manage schema version history and tenant access (in the case of a multi-tenant supported configuration). The schema contains 3 tables:

| Object | Notes | Purpose |
| ----- | ----- | ------- |
| VERSION_HISTORY | | Table used to track the schema version number for each of the managed objects in the data schema (usually FHIRDATA) |
| TENANTS | Db2 only | Table for mapping between tenant name and the tenant id used to isolate the tenant's data in the data schema |
| TENANT_KEYS | Db2 only | Authorization keys allocated to a given given. A tenant can have multiple valid keys at any point in time to support rolling key changes (add new before remove old).
| TENANT_SEQUENCE | Db2 only | Sequence used to provide unique mt_id values for allocated tenants |
| SET_TENANT | Db2 only | Stored procedure. The only way for FHIRUSER to set the value of the SV_TENANT_ID session variable.

These table definitions are more completely described in [DB2MultiTenancy.md](DB2MultiTenancy.md). 

----------------------------------------------------------------

# Schema: FHIR Resource Data 

> The IBM FHIR Server persists resources using tables defined in a data schema (usually FHIRDATA, but this can be anything). Multiple data schemas can exist in the same database, all managed from a single FHIR_ADMIN schema.

Reviewer: ```suggestion
The IBM FHIR Server persists resources using tables defined in a data schema (usually FHIRDATA). Multiple data schemas can exist in the same database, all managed from a single FHIR_ADMIN schema.
```

> > ![resource_schema.png](resource_schema.png)

Reviewer: I suggest removing the red squiggly lines
